hdl/interfaces/SPI.bsv

```diff
@@ -0,0 +1,747 @@
+// Copyright 2022 Oxide Computer Company
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package SPI;
+
+// BSV-provided
+import Clocks::*;
+import ClientServer::*;
+import Connectable::*;
+import FIFO::*;
+import FIFOF::*;
+import GetPut::*;
+import StmtFSM::*;
+import Vector::*;
+
+// Cobalt modules
+import RegCommon::*;
+
+interface SpiDecodeIF;
+    interface Server#(SpiRx, Bit#(8)) spi_byte;
+    interface Client#(RegRequest#(16, 8), RegResp#(8)) reg_con;
+endinterface
+
+
+typedef struct {
+    Maybe#(Bit#(8)) spi_rx_byte;
+    Bool done;
+} SpiRx deriving (Bits, Eq);
+
+typedef enum {
+    CMD, 
+    ADDR1, 
+    ADDR2, 
+    DO_READ, 
+    DO_WRITE, 
+    READ_WAIT, 
+    WRITE_WAIT
+} State deriving (Eq, Bits);
+
+//
+// This module provides a Client/Server interface
+// That takes byte-wise SPI payloads over its Server interface,
+// decodes the multi-byte spi protocol and issues Client transactions
+// to the register block and recieves responses from the register block.
+//
+// The SPI protocol looks like this:
+// <1byte Opcode> <1byte AddrH> <1byte AddrL> <n_bytes DATA>
+//
+module mkSpiRegDecode(SpiDecodeIF);
+    // Registers
+    Reg#(State) state <- mkReg(CMD);
+    Reg#(Bit#(16)) address <- mkRegU();
+    Reg#(RegOps) operation <- mkRegU();
+    Reg#(Maybe#(Bit#(8))) reg_read_data <- mkReg(tagged Invalid);
+    Reg#(Maybe#(Bit#(8))) reg_write_data <- mkReg(tagged Invalid);
+
+    // comb signals
+    RWire#(Bit#(8)) data <- mkRWire();
+    PulseWire spi_deselected <- mkPulseWire();
+    PulseWire got_data <- mkPulseWire();
+    let my_data = fromMaybe(?, data.wget());
+
+    // Store first byte which is the opcode
+    rule store_op (state == CMD);
```

Comment:
Kind of a nit, but if the first byte is the opcode should we name this state `OPCODE` instead of `CMD`? I don't necessarily care which we choose, but I think the comments and state should refer to it in the same way rather than command/opcode interchangeably.

Reply:
fixed